File: auto_permissions/providers/ollama.py
"""Ollama provider connector using /api/chat with discrete system and user roles."""
import json
import urllib.request
from typing import Any, Dict, Optional

from .base import BaseProvider, parse_json_safely

class OllamaProvider(BaseProvider):
    """Connects to Ollama using /api/chat with discrete system and user roles."""
    def __init__(
        self,
        endpoint: str = "http://127.0.0.1:11434/api/chat",
        model: str = "qwen3.5:9b",
        num_ctx: int = 4096,
        temperature: float = 0.0,
        timeout: float = 3.5,
    ):
        self.endpoint = endpoint.replace("localhost", "127.0.0.1")
        if self.endpoint.endswith("/api/generate"):
            self.endpoint = self.endpoint.replace("/api/generate", "/api/chat")
        self.model = model
        self.num_ctx = num_ctx
        self.temperature = temperature
        self.timeout = timeout

        self._cached_model_id: Optional[str] = None

    def _resolve_model_id(self) -> str:
        """Resolve Ollama model tag; queries /api/tags if model is 'auto' or 'default'."""
        if self.model and self.model not in ("auto", "default"):
            return self.model

        if self._cached_model_id:
            return self._cached_model_id

        tags_endpoint = self.endpoint.replace("/api/chat", "/api/tags")
        req = urllib.request.Request(tags_endpoint, headers={"Content-Type": "application/json"})
        try:
            with urllib.request.urlopen(req, timeout=1.0) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                models = [m.get("name", "") for m in data.get("models", []) if m.get("name")]
                if models:
                    # Preference order for Auto Permissions Mode
                    for pref in ("qwen3.5:9b", "qwen2.5:7b", "gemma4:e4b", "gemma4:12b", "gemma4:e2b"):
                        for m in models:
                            if m == pref or m.startswith(pref.split(":")[0]):
                                self._cached_model_id = m
                                return m
                    self._cached_model_id = models[0]
                    return models[0]
        except Exception:
            pass

        return "qwen3.5:9b"

    def evaluate(self, system_prompt: str, prompt: str) -> Optional[Dict[str, Any]]:
        resolved_model = self._resolve_model_id()
        payload = {
            "model": resolved_model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt}
            ],
            "format": "json",
            "stream": False,
            "options": {
                "num_ctx": self.num_ctx,
                "num_predict": 160,
                "temperature": self.temperature,
            }
        }
        req = urllib.request.Request(
            self.endpoint,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"}
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                content = data.get("message", {}).get("content") or data.get("response", "")
                return parse_json_safely(content)
        except Exception:
            return None

File: internal/providers/ollama.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

// OllamaProvider connects to Ollama /api/chat.
type OllamaProvider struct {
	Endpoint        string
	Model           string
	NumCtx          int
	Temperature     float64
	Timeout         time.Duration
	MaxTokens       int
	cachedModelID   string
	cachedModelTime time.Time
	mu              sync.Mutex
	client          *http.Client
}

func NewOllamaProvider(endpoint, model string, numCtx int, temp float64, timeout time.Duration, maxTokens int) *OllamaProvider {
	if endpoint == "" {
		endpoint = "http://127.0.0.1:11434/api/chat"
	}
	endpoint = strings.ReplaceAll(endpoint, "localhost", "127.0.0.1")
	if strings.HasSuffix(endpoint, "/api/generate") {
		endpoint = strings.ReplaceAll(endpoint, "/api/generate", "/api/chat")
	}
	if numCtx <= 0 {
		numCtx = 4096
	}
	if maxTokens <= 0 {
		maxTokens = 512
	}
	if timeout <= 0 {
		timeout = 3500 * time.Millisecond
	}
	return &OllamaProvider{
		Endpoint:    endpoint,
		Model:       model,
		NumCtx:      numCtx,
		Temperature: temp,
		Timeout:     timeout,
		MaxTokens:   maxTokens,
		client:      &http.Client{Timeout: timeout},
	}
}

func (p *OllamaProvider) GetEndpoint() string {
	return p.Endpoint
}

func (p *OllamaProvider) ResolveModelID() string {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.Model != "" && p.Model != "auto" && p.Model != "default" {
		return p.Model
	}

	if p.cachedModelID != "" && time.Since(p.cachedModelTime) < 60*time.Second {
		return p.cachedModelID
	}

	tagsURL := strings.ReplaceAll(p.Endpoint, "/api/chat", "/api/tags")
	client := &http.Client{Timeout: 1000 * time.Millisecond}
	req, err := http.NewRequest("GET", tagsURL, nil)
	if err == nil {
		req.Header.Set("Content-Type", "application/json")
		if resp, err := client.Do(req); err == nil {
			defer resp.Body.Close()
			if resp.StatusCode == http.StatusOK {
				body, _ := io.ReadAll(resp.Body)
				var data struct {
					Models []struct {
						Name string `json:"name"`
					} `json:"models"`
				}
				if err := json.Unmarshal(body, &data); err == nil && len(data.Models) > 0 {
					var names []string
					for _, m := range data.Models {
						if m.Name != "" {
							names = append(names, m.Name)
						}
					}
					for _, pref := range []string{"qwen3.5:9b", "qwen2.5:7b", "gemma4:e4b", "gemma4:12b", "gemma4:e2b"} {
						prefBase := strings.Split(pref, ":")[0]
						for _, n := range names {
							if n == pref || strings.HasPrefix(n, prefBase) {
								p.cachedModelID = n
								p.cachedModelTime = time.Now()
								return n
							}
						}
					}
					if len(names) > 0 {
						p.cachedModelID = names[0]
						p.cachedModelTime = time.Now()
						return names[0]
					}
				}
			}
		}
	}

	return "qwen3.5:9b"
}

func (p *OllamaProvider) Evaluate(systemPrompt, prompt string) (map[string]interface{}, string, error) {
	resolvedModel := p.ResolveModelID()
	payload := map[string]interface{}{
		"model": resolvedModel,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": prompt},
		},
		"format": "json",
		"stream": false,
		"options": map[string]interface{}{
			"num_ctx":     p.NumCtx,
			"num_predict": p.MaxTokens,
			"temperature": p.Temperature,
		},
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return nil, "LOCAL", err
	}

	req, err := http.NewRequest("POST", p.Endpoint, bytes.NewReader(data))
	if err != nil {
		return nil, "LOCAL", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, "LOCAL", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, "LOCAL", fmt.Errorf("ollama returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, "LOCAL", err
	}

	var dataResp map[string]interface{}
	if err := json.Unmarshal(body, &dataResp); err != nil {
		return nil, "LOCAL", err
	}

	var content string
	if msg, ok := dataResp["message"].(map[string]interface{}); ok {
		content, _ = msg["content"].(string)
	}
	if content == "" {
		content, _ = dataResp["response"].(string)
	}

	parsed := ParseJSONSafely(content)
	if parsed == nil {
		return nil, "LOCAL", fmt.Errorf("failed to parse ollama json response")
	}
	return parsed, "LOCAL", nil
}
